uniffi_core/src/ffi/rustfuture.rs

```diff
@@ -286,36 +287,41 @@ where
         };
     }
 
-    fn make_waker(self: &Pin<Arc<Self>>) -> Waker {
+    fn make_waker(self: &Arc<Self>) -> Waker {
         // This is safe as long as we implement the waker interface correctly.
         unsafe {
             Waker::from_raw(RawWaker::new(
+                // `self` is cloned, and `Self::into_raw` is called immediately. It could leak
+                // memory if the `Waker` doesn't call `Self::from_raw`.
                 self.clone().into_raw(),
                 &Self::RAW_WAKER_VTABLE,
             ))
         }
     }
 
+    /// Cancel or drop the `Future` contained inside `Self`.
+    pub fn cancel_or_drop_inner_future(self: Arc<Self>) {
+        // Increase the `wake_counter` to ensure the future cannot be awaken again.
+        self.wake_counter.fetch_add(1, Ordering::Relaxed);
+
+        // Take the `Future`, leaving `None` in its place.
+        let _future = self.future.write().unwrap().take();
```

Comment:
This ensures `self.future` is dropped, but what about the rest of `self`?  Are we sure that the `Arc<Self>` will eventually be dropped.

One scenario I'm thinking about is if there's already a `wake_callback` scheduled.  In that case, we've called `Arc::into_raw()` with the expectation of calling `Arc::from_raw()` later, but that's not going to happen.  It seems dirty, but maybe if `wake_counter` wasn't 0 before the `fetch_add` call, we should manually decrement the count.

Reply:
Thinking about this a bit more, it could be a tricky issue.  On Kotlin, it seems that the scheduled callbacks are never called after a `CoroutineScope` is cancelled, but I don't know if that's true for all the foreign bindings.

Reply:
> Thinking about this a bit more, it could be a tricky issue. On Kotlin, it seems that the scheduled callbacks are never called after a `CoroutineScope` is cancelled, but I don't know if that's true for all the foreign bindings.

Yeah, the way Kotlin handles that is really not simple to fit in the current design of `ForeignExecutor`.